### ISSUE TITLE
[NVIDIA Proprietary] [arch/artix] kwin_x11 compositor extremely laggy after hijacking arch/artix linux

### ISSUE THREAD

**Issue by A-Human-Person** [bedrocklinux/bedrocklinux-userland#140]
similar to the long lasting bug on nvidia proprietary before it was fixed, this is almost certainly a proprietary driver exclusive issue. this does not happen with mutter so if this ends up being a plasma issue (it likely is) I will file a bug report there.

**Comment by paradigm** [comment]
To make sure I understand correctly:

- You install Arch/Artix and set things up to your liking and the kwin_x11 compositor is fine.
- You run the Bedrock Linux installer script.
- You reboot.
- With no other changes, the kwin_x11 compositor is now laggy.

Is that correct?

If so, that's very strange.  At that point, Bedrock has yet to really touch anything related to the graphics stack.  It should be the same kernel, the same kernel modules, the same userland drivers, and the same compositor.  The only possibility that comes to my mind at the moment is that part of kwin_x11 is tripping on something Bedrock does distant from the graphics stack, like `etcfs` or `crossfs`.  If you load up `htop` or some other processor monitor, are one or more CPUs pegged?  If so, can you tell what the processes are?

**Comment by A-Human-Person** [comment]
yes,
installed arch/artix
installed the packages from plasma
started plasma from sddm
ran script
sudo systemctl reboot -i/power button
kwin is laggy now


and yes, I see atleast 12 etcfs tasks towards the top of htop. it doesnt seem to be taking many resources though. kwin is taking up ~100-150mb of ram at all time (should be normal)

**Comment by A-Human-Person** [comment]
forgot to mention I also installed the nvidia package for the prop driver

**Comment by paradigm** [comment]
I'm less interested in the process count and memory usage then CPU usage.  Once the system is hijacked and reproducing the laggy environment, can you run:

```
top -bn1 > /tmp/top-log
```

and provide the contents of `/tmp/log`?  That should catch the CPU utilization I'm interested in.

Also run:

```
brl report /tmp/brl-report-log
```

and provide the contents of `/tmp/brl-report-log`.  While doubtful, maybe Bedrock's health check will see some issue.

If these do not illustrate any problems, I'm at a loss for what could be going on.  I don't see how Bedrock's changes could make a difference here, nor do I see how it could be a bug in the nVidia drivers.

I can probably get my hands on a physical box with an nVidia card that I can test install to reproduce the issue at some point, but may be a bit before I get around to it.

**Comment by A-Human-Person** [comment]
ok im gonna join irc

**Comment by paradigm** [comment]
For anyone else who runs into this thread, the IRC conversation included what appeared to be indications that etcfs was getting spammed.  A suspicion was that kwin_x11 was spamming `/etc/localtime` to get the date, but a quick and dirty test of bind mounting a `/usr/share/zoneinfo/...` file to `/etc/localtime` didn't resolve it, so there's something else going on.

My plan going forward is to try to reproduce it locally.  If I can't do that, I'll look into making a debug `etcfs` build which could highlight the concern.

**Comment by paradigm** [comment]
I ran Arch's `kwin_x11` locally with proprietary nVidia drivers.  I was able to reproduce `kwin_x11` making an excessive number of calls to `stat()` on `/etc/localtime`.  However, I wasn't able to reproduce any resulting lag.  It's possible this is because we're going off in the wrong direction.  `kwin_x11` also appears to make an excessive number of calls to `getpid()` and other things.  However, it's also possible that it's just a performance difference between our boxes, and we're on the right track.

@A-Human-Person I asked you to run a test on IRC.  It's possible the test wasn't run at the right time/place.  Lets try it again, with a tweak.  On the machine which reproduces the issue, can you edit `/bedrock/strata/bedrock/sbin/init` as root.  There's a blank line under

```
step "Handing control off to ${color_strat}${init_stratum}${color_glue}:${color_cmd}${init_cmd}${color_norm}"
```

[here](https://github.com/bedrocklinux/bedrocklinux-userland/blob/6d96ea230c5475837898d53305a2bf8132142e7b/src/init/init#L430).

In that blank line, add:

```
mount --bind /usr/share/zoneinfo/UTC /etc/localtime
```

so that with context it looks like:

```
step "Handing control off to ${color_strat}${init_stratum}${color_glue}:${color_cmd}${init_cmd}${color_norm}"
mount --bind /usr/share/zoneinfo/UTC /etc/localtime
if ! [ -x "${init_cmd}" ]; then
```

Then reboot.  Provided your only strata are Bedrock and Arch/Artix, this should make it so `kwin_x11`'s `stat("/etc/localtime",...)` calls go directly to a timezone file instead of `etcfs`.

With this change in place, I'd like to check two things.

First, does it continue to reproduce the apparent legginess, or is that resolved?

Second, use the system for a while.  Comparable to however long you've used the system previously before providing

```
top -bn1 > /tmp/top-log
```

for me earlier.  Once that's the case, run

```
top -bn1 > /tmp/top-log
```

and provide the new `/tmp/top-log` contents.  We'll see if the `mount --bind` change described above actually resolves the `etcfs` CPU usage we saw earlier.

**Comment by A-Human-Person** [comment]
ok so, it appears the lag is much less dramatic, but there is still lag. almost exactly like the nvidia bug that was recently fixed.

`top -bn1 > /tmp/top-log` doesn't appear to do anything
manually editing the file with vim or nano gives

```  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 2763 root      20   0    3012    976     68 S   5.9   0.0   0:47.89 etcfs
 6809 spencer   20   0 2956432 162224  91680 S   5.9   1.0   0:38.42 kwin_x11
 7945 spencer   20   0  399036 164420 102872 S   5.9   1.0   0:15.69 steam
 9511 spencer   20   0  353644  68468  56088 S   5.9   0.4   0:00.92 konsole
 9674 spencer   20   0   13264   3732   3108 R   5.9   0.0   0:00.02 top
```

showing that this apparently did not fix the problem.

**Comment by A-Human-Person** [comment]
also I just force killed etcfs and all the lag completely dissapeared instantly

**Comment by paradigm** [comment]
I would caution against using Bedrock without `etcfs`; things could get really wonky.  However, that's certainly a definitive test to indicate the issue is in fact etcfs related.

Next step would be to add debug code to `etcfs` for you to try out and report back with.  As I'm currently imagining it, I could probably release the functionality in a beta with it defaulting to off such that all you would have to do is enable it in `bedrock.conf` and provide the resulting logs.

**Comment by A-Human-Person** [comment]
yeah it broke a ton of things killing it but it definitely showed it was the problem

**Comment by A-Human-Person** [comment]
going to join irc when I have the chance to test this. will also do a fresh install since I have a bit of stuff installed.

**Comment by paradigm** [comment]
@A-Human-Person I pushed 0.7.8beta2 with debug code.  [Update to it](https://bedrocklinux.org/0.7/beta-channel.html) then edit your `/bedrock/etc/bedrock.conf` and add `etcfs` to the `debug =` line under `[miscellaneous]`.  With context it'll look something like:

```
[miscellaneous]

#
# Time to retain Bedrock cached data (such as from brl-fetch) in days.
#
cache-life = 90

#
# Set to false to disable color in brl output
#
color = true

#
# Enable debugging for specified subsystems.
#
debug = etcfs
```

Then reboot.  Bedrock is now logging `etcfs` calls.  Use the system for a bit, reproducing the laggy environment issue.

Once you're satisfied that the lagginess should be properly captured in the `etcfs` logs, edit your `bedrock.conf` and remove the newly added `etcfs` value for `[miscellaneous]/debug`.  Then reboot again to stop the logging.

You should find `/bedrock/var/cache/etcfs-*/log` files.  They might be too large for things like hastebin.  However, they _should_ compress well.  See if you can maybe compress them, e.g. `zip` or `gzip` or `xz` or something, then make them available to me.  I'll go through them and see if it helps illuminate what the issue is.

**Comment by A-Human-Person** [comment]
m

**Comment by A-Human-Person** [comment]
[logs.tar.gz](https://github.com/bedrocklinux/bedrocklinux-userland/files/3598473/logs.tar.gz)

**Comment by paradigm** [comment]
That log caught 113517 requests to Arch's etcfs. 74727 of them, about 66%, were about `/etc/localtime`.  These occurred over 208 seconds, which makes it 359 requests about `/etc/localtime` per second.

Many of those requests were just one time items initializing things, such as reading font configuation.  The bulk of repeating requests were from `/usr/bin/plasmashell` (28182) and `/usr/bin/kwin_x11` (35406) about `/etc/localtime`.  That's about 85% of the `/etc/localtime` requests.

This does very much seem like the culprit.

It wasn't originally clear to me why the `mount --bind` strategy I proposed earlier didn't resolve the issue.  Apparently `mount` follows symlinks:

```
# readlink /etc/localtime
/bedrock/run/localtime
# mount --bind /bedrock/cross/zoneinfo/America/New_York /etc/localtime
# readlink /etc/localtime
/bedrock/run/localtime
# umount /bedrock/cross/zoneinfo/America/New_York
# echo $?
0
```

That's my bad, I feel silly now.

This is arguably a good thing - I can continue to pursue this strategy to resolve the issue.  My current plan here is to completely rework the timezone stuff to have etcfs force `/etc/localtime` to be a normal, non-symlink file, and then have `brl enable` bind mount the specified timezone over it.  There are some potential concerns here, such as `systemd` trying to set the timezone and either getting upset at the resulting error or changing the actual underlying timezone file.  I'll have to experiment with it.  I'll see if I can queue it up for beta3.

Many thanks for your patience gathering debug info here.

**Comment by A-Human-Person** [comment]
you don’t have to apologize, this is probably one of the only issues in bedrock right now. I am not directly bothered by this since compiz works. surprised PlasmaShell does this too. I’m amazed you figured out the problem within a minute without any debug info then the debug info proved you to be correct. thanks paradigm.

**Comment by A-Human-Person** [comment]
also I find it weird that kwin/plasmashell doesn’t have a limit on /etc/localtime requests, that seems like an issue with plasma itself. If I knew C(++) I would make a pull request that limits requests since that shouldn’t be happening in the first place.

**Comment by paradigm** [comment]
Agreed.  Over three hundred requests a second certainly seems excessive.  It could be worth making a bug report over.

Irrelevant of whether or not it's resolved on KDE's end, I'd like to see if I can get Bedrock to handle the situation better itself.  If the KDE folks can run into this kind of issues, presumably so can others.  From most software consumer's point of view, it doesn't really matter whose fault it is; if it doesn't happen with their previous distro, but it does with Bedrock, it's effectively a problem with Bedrock.

**Comment by A-Human-Person** [comment]
yeah, but this seems to be a common issue is plasma and this also happened with nvidia proprietary before it was fixed for **years**. Every check should be limited to 1 check every 5 seconds unless debug settings are on.

**Comment by paradigm** [comment]
Thinking about it more, I'm increasingly concerned about systemd following a potential bind mount to a timezone file and then changing that file.  I might pursue another strategy before the bind mount one, if you don't mind a potential extra round of testing.

Currently, when KDE makes a request about `/etc/localtime`:

- etcfs then triggers a sanity check to ensure the underlying `/etc/localtime` is still the Bedrock enforced symlink
- etcfs then forwards the underlying `/etc/localtime` symlink
- KDE follows the symlink to `/bedrock/run/localtime`, another symlink
- KDE follows the symlink to `/bedrock/cross/zoneinfo/<timezone>`, another Bedrock FUSE filesystem
- crossfs searches through the strata for the highest priority stratum that provides such a timezone file
- crossfs returns the underlying file

On my system, all that together takes:

```
$ time cat /etc/localtime >/dev/null
cat /etc/localtime > /dev/null  0.00s user 0.00s system 0% cpu 0.003 total
```

a very small amount of time, which is why I previously deemed it acceptable.  However, doing it 360 times a second:

```
$ time sh -c 'for x in $(seq 1 360); do cat /etc/localtime >/dev/null; done'
sh -c 'for x in $(seq 1 360); do cat /etc/localtime >/dev/null; done'  0.00s user 0.02s system 3% cpu 0.531 total
```

becomes problematic.  I might rework things thusly:

- Whenever a stratum is enabled or disabled, all of the etcfs instances on the system have cached localtime contents updated.
- Whenever anyone queries etcfs about /etc/localtime, etcfs returns cached file contents.

No symlinks, no bind mounts.  No way for something to follow any redirections from `/etc/localtime` to a file I don't want them to change and then change it.

It'll make enabling/disabling strata a touch slower, but I think that's acceptable, as it happens pretty rarely.  etcfs has fundamental overhead, and so the runtime part is also slower than the theoretical ideal, but should be a lot faster than the current pattern.  I'm hoping its fast enough.

**Comment by paradigm** [comment]
I might have found a best of both worlds solution: read-only bind mounts.  This has the performance of a bind mount, completely bypassing etcfs, but with the restriction that systemd isn't going to be able to change it or the underlying file through it.

```
# echo "original-content" > backing-file
# touch link
# mount --bind backing-file link
# mount -o remount,bind,ro backing-file link
# cat backing-file link
original-content
original-content
# echo "new-content" | tee link >/dev/null
tee: link: Read-only file system
# cat backing-file link
original-content
original-content
# echo "new-content" | tee backing-file >/dev/null
# cat backing-file link
new-content
new-content
```

The main concern here is that systemd will _know_ that it can't change the file and become upset.  If that's the case, this strategy isn't viable, and I'll have to revert to just optimizing etcfs as much as possible for `/etc/localtime` handling.  I can't avoid the FUSE overhead then, but I can probably avoid making any other system calls.

**Comment by paradigm** [comment]
I tested upgrading Debian from Stretch to Buster with the `/etc/localtime` bind mount strategy, and it did not go well:

```
Setting up dzdata (2019b-0+deb9u1)
mv: cannot move '/etc/localtime.dpkg.new' to '/etc/localtime': Device or resource busy
dpkg: error processing package tzdata (--configure):
  subprocess installed post-installation script returned error exit status 1
```

The only tools left in my tool belt here are:

- General etcfs performance improvements.
	- There's a libfuse feature called "readdirplus" that did not work as expected last time I looked into it.  I should revisit it.
	- Maybe we can leverage caching, e.g. https://github.com/libfuse/libfuse/issues/394 or https://github.com/libfuse/libfuse/issues/362
- Heavily special casing `/etc/localtime` handling in etcfs.  Have every bit of code that _could_ handle `/etc/localtime` earlier do so.

Hopefully those are enough to resolve the concern here.  If not, we may have to go dig into why KDE is querying `/etc/localtime` so often and potentially throw patches at them.

**Comment by A-Human-Person** [comment]
I think the best option is to fix it on kde's end, afaik no other window managers/shells do this. I can look at the code and see if I can figure it out but I'm not fluent in any languages right now.

**Comment by A-Human-Person** [comment]
actually this appears to not be the only sketchy thing that kwin does, https://github.com/tildearrow/kwin-lowlatency

**Comment by paradigm** [comment]
It's an easy bug to accidentally create.  A sleep deprived programmer might try to avoid querying for the time too often by checking the time since the last time time was checked.  I think it not unlikely other software out there does something similar, and so irrelevant of specifically KDE I think it worthwhile for Bedrock to improve performance here as much as is feasible without major sacrifices elsewhere.

That having been said, fixing it in KDE as well would certainly be ideal.  You could bug report it there, if nothing else; maybe they'll fix it on their end without a patch.  If you do have the time and skill to offer a patch, my guess is whatever they're doing only needs the _relative_ time passed rather than the absolute local time.  If that's the case, [they could use `CLOCK_MONOTONIC`](https://linux.die.net/man/3/clock_gettime) which I would expect does not read `/etc/localtime` at all.

**Comment by A-Human-Person** [comment]
thats what I planned, thanks.

**Comment by parkerlreed** [comment]
Also running into this.

Video: https://www.youtube.com/watch?v=ZBeXYFe3ltI

I'll take some time to dig in and see if I can't poke a few devs.

**Comment by Human1895** [comment]
yeah thanks for looking into this, if you read above at my horrible debugging from last year basically /etc/localtime goes through a ton of symlinks and it takes a ton of processing power which can be seen by looking at the etcfs process, kde devs would need to switch their method of getting the time.

interesting that you also have the same issue as me, and with different hardware, it looked exactly the same on your screen as when I used bedrock/nvidia. I’d expect this to be an issue for everyone and not just a few people. I haven’t used bedrock with my amd gpu so I don’t know if it persisted or not after switching.

**Comment by paradigm** [comment]
> if you read above at my horrible debugging from last year basically /etc/localtime goes through a ton of symlinks and it takes a ton of processing power which can be seen by looking at the etcfs process

To be clear, it's not actually the symlinks that's the problem, but FUSE.  With some effort I can get rid of the symlinks, but not FUSE.

When you typically do something with/to a file on a Linux distro, you're sending a request to the Linux kernel which implements stuff like ext4 and vfat and btrfs.  On Bedrock, when you do anything with or in or to `/etc`, after you ask the Linux kernel for whatever `/etc` operation you're interested in, the kernel then forwards the request it to a Bedrock program called `etcfs`.  This gives Bedrock a lot of control over `/etc` and is essentially for some of what Bedrock does.  However, the interface between the kernel and `etcfs`, called `FUSE`, is somewhat slow.  I can in theory make everything around `FUSE` fast, including removing all the symlinks, but the issue here will continue to persist.  There are efforts to speed up or replace FUSE, such as [ExtFUSE](https://www.phoronix.com/scan.php?page=news_item&px=ExtFUSE-Faster-FUSE-eBPF), which I'd love to move over to if/when they become available.

I fully intend to getting around to optimizing for `/etc/localtime` performance at some point.  I can tweak things in `etcfs` so it handles `/etc/localtime` before the general case (making `/etc/localtime` ever so slightly faster, but everything else slower).  I can compare having `etcfs` cache the underlying `localtime` contents in RAM and provide them immediately against a symlink (which might be faster because it's less content that has to be sent through FUSE).  However:

- This isn't a trivial change, and I don't want to rush it and risk bugs.
- I don't have much confidence it'll actually help very much.  Because of how FUSE works, it's overhead will still be enough to cause problems when called hundreds of times per second synchronously with a UI element.
- I have a lot of other competing things people very badly want me to do.

It'll be a while before I can get to it.

> kde devs would need to switch their method of getting the time.

This would be ideal.  I suspect it should be quite possible to fix on their end.  I don't think their users want it stat'ing `/etc/localtime` hundreds of times per second even on systems where it is performant.  I haven't personally gone to the KDE folks, as both of you - Human and Parker - gave me the impression you might.  **If I misunderstood and that's not the case, let me know and I'll look into making a bug report and/or upstreaming a patch.**

> I’d expect this to be an issue for everyone and not just a few people.

Agreed, I expect the fundamental issue should hit people consistently on Bedrock+KDE.  I think the graphics hardware/driver is a red herring here.  I reproduced the underlying issue as well, I just didn't recognize the resulting performance lag.  Maybe my hardware is faster which hid the issue, or maybe I didn't have enough experience for what KDE is supposed to feel like without this issue.

-----

Parker did some digging and found that `TZ=:/etc/localtime` resolves the issue for him.  When `TZ` is set, glibc will skip `stat()`'ing `/etc/localtime` if queried too often.  However, it's not a perfect solution.

I was aware of the general performance difference with `TZ` unset and with `TZ=:/etc/localtime` and so Bedrock _used_ to set `TZ=:/etc/localtime`.  [However, `TZ:=/etc/localtime` triggers a bug in Chromium (and a huge derivative ecosystem, including chromium, steam, discord, signal, etc)](https://bugs.chromium.org/p/chromium/issues/detail?id=811403).  I think Chromium implemented their own timezone logic rather than use libc's, and they didn't implement the `TZ` standard fully.  [People have tried to upstream a fix but the effort seems to have stalled](https://chromium-review.googlesource.com/c/chromium/deps/icu/+/1006219/).  I had a _lot_ of people bug report issues with Chromium-derived software on Bedrock, and so I went the other way and had Bedrock `unset TZ`, as that caused everything I tested to read `/etc/localtime` to get timezone information and "just worked."  IIRC I actually tested KDE when experimenting here, but was myopic about the clock working rather than A:B testing performance.

If you want to use KDE and don't mind the Chromium issue, you can switch back to Bedrock's previous TZ strategy by opening up `/bedrock/etc/brl-apply` and changing [this line](https://github.com/bedrocklinux/bedrocklinux-userland/blob/9c706ea96e1bb93e05d19965dbf869e04601d2b2/src/slash-bedrock/libexec/brl-apply#L132):

```
	echo "unset TZ"
```

to

```
	echo "TZ=:/etc/localtime"
```

If you have interest in `fish`, also change [this line](https://github.com/bedrocklinux/bedrocklinux-userland/blob/9c706ea96e1bb93e05d19965dbf869e04601d2b2/src/slash-bedrock/libexec/brl-apply#L151):

```
	echo "set -e TZ"
```

to

```
	echo "set TZ :/etc/localtime"
```

Then reboot.  You'll probably find KDE is much faster because KDE effectively queries `/etc/localtime` far less often.  Note this does not change Bedrock component performance; `/etc/localtime` is still a FUSE symlink.

The down-sides to this are:

- For the time being, a `brl update` will wipe the change, and you'll have to re-apply it every time.  I can resolve that by making this a `bedrock.conf` option.  I'll look into it when I find the time and note its existence here.
- If you run into an issue with a Chromium-derived application acting weird, try wrapping it with a script that `unset TZ`'s before calling the application.

-----

I'm stuck in a hard place between slower-than-ideal Bedrock code that I don't know how to technically fix, slower-than-ideal KDE code that is probably fixable if we can convince the KDE folks to agree to it, and non-standard-fulfilling Chromium code that people have tried to update but stalled.  Given that I don't know how to make Bedrock do its thing without FUSE, and that people have already tried to fix it in Chromium but it didn't take, it feels like the best strategy here is to fix it in KDE if we can.  I don't like shirking responsibility and typically prefer to fix things on Bedrock's end where possible, but I don't see way to do so in this instance.

**Comment by nift4** [comment]
Is this fixed by the commit above?

**Comment by paradigm** [comment]
Yes, I believe it is.  Thanks for reviewing curate the issue list!
